contracts/services/osmosis-cl-withdrawer/src/contract.rs

```diff
@@ -0,0 +1,161 @@
+    cfg: Config,
+    position_id: u64,
+    liquidity_amount: String,
+) -> Result<Response, ServiceError> {
+    // here we just assert that the position exists.
+    // any validations beyond this (like position ownership, etc.)
+    // will propagate on execution.
+    let _position_query_response: PositionByIdResponse = deps
+        .querier
+        .query(&PositionByIdRequest { position_id }.into())
+        .map_err(|_| StdError::generic_err("no such position"))?;
+
+    let liquidate_position_msg = MsgWithdrawPosition {
+        position_id,
+        sender: cfg.input_addr.to_string(),
+        liquidity_amount,
+    };
+
+    // we delegate the position liquidation msg as a submsg because we
+    // will need to transfer the underlying tokens we liquidate afterwards.
+    let delegated_input_acc_msgs = execute_submsgs_on_behalf_of(
+        vec![SubMsg::reply_on_success(liquidate_position_msg, REPLY_ID)],
+        Some(to_json_string(&cfg)?),
+        &cfg.input_addr.clone(),
+    )?;
+
+    let service_submsg = SubMsg::reply_on_success(delegated_input_acc_msgs, REPLY_ID);
+
+    Ok(Response::default().add_submessage(service_submsg))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::Ownership {} => {
+            to_json_binary(&valence_service_base::get_ownership(deps.storage)?)
+        }
+        QueryMsg::GetProcessor {} => {
+            to_json_binary(&valence_service_base::get_processor(deps.storage)?)
+        }
+        QueryMsg::GetServiceConfig {} => {
+            let config: Config = valence_service_base::load_config(deps.storage)?;
+            to_json_binary(&config)
+        }
+        QueryMsg::GetRawServiceConfig {} => {
+            let raw_config: ServiceConfig =
+                valence_service_utils::raw_config::query_raw_service_config(deps.storage)?;
+            to_json_binary(&raw_config)
+        }
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn reply(deps: DepsMut, _env: Env, msg: Reply) -> Result<Response, ServiceError> {
+    match msg.id {
+        REPLY_ID => handle_liquidity_withdrawal_reply(deps.as_ref(), msg.result),
+        _ => Err(ServiceError::Std(StdError::generic_err("unknown reply id"))),
+    }
+}
+
+fn handle_liquidity_withdrawal_reply(
+    deps: Deps,
+    result: SubMsgResult,
+) -> Result<Response, ServiceError> {
+    // load the config that was used during the initiating message
+    // which triggered this reply
+    let cfg: Config = parse_valence_payload(&result)?;
+
+    // decode the response from the submsg result
+    let valence_callback = ValenceCallback::try_from(result)?;
+
+    // decode the underlying position withdrawal response
+    // and query the pool to match the denoms
+    let decoded_resp: MsgWithdrawPositionResponse = valence_callback.result.try_into()?;
+    let pool = query_cl_pool(&deps, cfg.pool_id.u64())?;
```

Comment:
Can't we query the balances here? seems like a more reliable way of getting the amounts then to parse the reply

Reply:
we can but I think it's a bit extreme to just forward entire balances regardless of what happened, especially given that the same account could be used as input for multiple services

Reply:
Only suggested it because of the "rely on reply" fear, I try to never rely on any response we get from reply except the fact we got it, probably changed in Cosmwasm 2.0